near-sdk/src/private/abi.rs

```diff
@@ -0,0 +1,183 @@
+    #[serde(default)]
+    pub name: Option<String>,
+    /// The version of the smart contract.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub version: Option<String>,
+    /// The authors of the smart contract.
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    #[serde(default)]
+    pub authors: Vec<String>,
+    /// Other arbitrary metadata.
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub other: HashMap<String, String>,
+}
+
+/// Core ABI information.
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct Abi {
+    /// ABIs of all contract's functions.
+    pub functions: Vec<AbiFunction>,
+    /// Root JSON Schema containing all types referenced in the functions.
+    pub root_schema: RootSchema,
+}
+
+impl Abi {
+    pub fn combine<I: IntoIterator<Item = Abi>>(abis: I) -> Abi {
+        let mut functions = Vec::<AbiFunction>::new();
+        let mut gen = schemars::gen::SchemaGenerator::default();
+        let definitions = gen.definitions_mut();
+
+        for abi in abis {
+            // Update resulting JSON Schema
+            definitions.extend(abi.root_schema.definitions.to_owned());
+
+            // Update resulting function list
+            functions.extend(abi.functions);
+        }
+        // Sort the function list for readability
+        functions.sort_by(|x, y| x.name.cmp(&y.name));
+
+        Abi { functions, root_schema: gen.into_root_schema_for::<String>() }
+    }
+}
+
+/// ABI of a single function.
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct AbiFunction {
+    pub name: String,
+    /// Whether function does not modify the state.
+    #[serde(skip_serializing_if = "is_false")]
+    #[serde(default)]
+    pub is_view: bool,
+    /// Whether function can be used to initialize the state.
+    #[serde(skip_serializing_if = "is_false")]
+    #[serde(default)]
+    pub is_init: bool,
+    /// Whether function is accepting $NEAR.
+    #[serde(skip_serializing_if = "is_false")]
+    #[serde(default)]
+    pub is_payable: bool,
+    /// Whether function can only accept calls from self (current account).
+    #[serde(skip_serializing_if = "is_false")]
+    #[serde(default)]
+    pub is_private: bool,
+    /// Type identifiers of the function parameters.
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    #[serde(default)]
+    pub params: Vec<AbiParameter>,
+    /// Type identifiers of the callbacks of the function.
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    #[serde(default)]
+    pub callbacks: Vec<AbiType>,
+    /// Type identifier of the vararg callbacks of the function.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub callbacks_vec: Option<AbiType>,
+    /// Return type identifier.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub result: Option<AbiType>,
```

Comment:
Same as previous comment, serde attributes can be combined to one line if you want